### DISCUSSION TITLE
strongswan error plugin description of errors

### DISCUSSION THREAD

**Discussion by amodster** [strongswan/strongswan#1681] [General]
Hello
I am looking for description information on error plugins messages. What causes the error and how does charon handle the error (will it continue to retry, etc..

We are using strongswan-5.8.2, and looking at file error_notify_msg.h and error types below:

enum {
	ERROR_NOTIFY_RADIUS_NOT_RESPONDING = 1,
	ERROR_NOTIFY_LOCAL_AUTH_FAILED = 2,
	ERROR_NOTIFY_PEER_AUTH_FAILED = 3,
	ERROR_NOTIFY_PARSE_ERROR_HEADER = 4,
	ERROR_NOTIFY_PARSE_ERROR_BODY = 5,
	ERROR_NOTIFY_RETRANSMIT_SEND_TIMEOUT = 6,
	ERROR_NOTIFY_HALF_OPEN_TIMEOUT = 7,
	ERROR_NOTIFY_PROPOSAL_MISMATCH_IKE = 8,
	ERROR_NOTIFY_PROPOSAL_MISMATCH_CHILD = 9,
	ERROR_NOTIFY_TS_MISMATCH = 10,
	ERROR_NOTIFY_INSTALL_CHILD_SA_FAILED = 11,
	ERROR_NOTIFY_INSTALL_CHILD_POLICY_FAILED = 12,
	ERROR_NOTIFY_UNIQUE_REPLACE = 13,
	ERROR_NOTIFY_UNIQUE_KEEP = 14,
	ERROR_NOTIFY_VIP_FAILURE = 15,
	ERROR_NOTIFY_AUTHORIZATION_FAILED = 16,
	ERROR_NOTIFY_CERT_EXPIRED = 17,
	ERROR_NOTIFY_CERT_REVOKED = 18,
	ERROR_NOTIFY_NO_ISSUER_CERT = 19,
	ERROR_NOTIFY_RETRANSMIT_SEND = 20,
};

**Comment by tobiasbrunner** [comment]
> I am looking for description information on error plugins messages.

The listener that maps the alerts to these values also generates a text message for each error, see [here](https://github.com/strongswan/strongswan/blob/master/src/libcharon/plugins/error_notify/error_notify_listener.c#L60).

> What causes the error and how does charon handle the error (will it continue to retry, etc..

Whether there is a recovery depends on the error and in some cases the configuration. You could search for `ALERT_` in the source code to see where a particular alert/error is triggered and see what happens.

**Comment by amodster** [comment]
Tobias
I am trying to find the caller to one of the IKE authentication functions.
I am losing track at the function source file (there should be a macro define somewhere) ?

For example:
ERROR_NOTIFY_LOCAL_AUTH_FAILED

were using ikev2
METHOD(task_t, build_i, status_t,
                private_ike_auth_t *this, message_t *message)
src/libcharon/sa/ikev2/tasks/ike_auth.c:                                              charon->bus->alert(charon->bus, ALERT_LOCAL_AUTH_FAILED);
src/libcharon/sa/ikev2/tasks/ike_auth.c:                                              charon->bus->alert(charon->bus, ALERT_LOCAL_AUTH_FAILED);
src/libcharon/sa/ikev2/tasks/ike_auth.c:                                              charon->bus->alert(charon->bus, ALERT_LOCAL_AUTH_FAILED);

# sa/ikev2/tasks/ike_auth.lo - a libtool object file
# Generated by ltmain.sh (GNU libtool) 2.2.4
#
# Please DO NOT delete this file!
# It is necessary for linking the library.

# Name of the PIC object.
pic_object='.libs/ike_auth.o'

# Name of the non-PIC object
non_pic_object=none

nm .libs/ike_auth.o

nm .libs/ike_auth.o
00001350 t T.155
00000ab0 t add_ppk_identity
00000758 t apply_auth_cfg
00001eb8 t apply_ppk
         U auth_cfg_create
         U authenticator_create_builder
         U authenticator_create_verifier
00000be0 t build_i
00001f88 t build_r
         U charon
         U chunk_create_cat
         U chunk_create_clone
         U chunk_empty
         U chunk_length
00001e38 t clear_ppk
00000000 t collect_my_init_data
00000098 t collect_other_init_data
         U delete_ike_sa_job_create
000029b8 t destroy
00000220 t do_another_auth
         U free
00000128 t get_auth_cfg
00000938 t get_ppk
000006d0 t get_reserved_id_bytes
000005d0 t get_type
         U id_payload_create_from_identification
         U identification_create_from_data
         U identification_create_from_encoding
         U identification_create_from_sockaddr
000005d8 T ike_auth_create
         U lib
         U linked_list_create
         U malloc
         U memset
         U memwipe_noinline
         U message_create
00002850 t migrate
         U notify_type_names
         U payload_get_field
00002ae0 t process_i
000013f8 t process_r
         U redirect_data_create
         U redirect_data_parse
000007f0 t send_auth_failed_informational
00000350 t update_cfg_candidates

I should see the METHOD() in the map above ?

Thanks

**Reply by tobiasbrunner** [reply]
> I should see the METHOD() in the map above ?

You see the `build_i` symbol in which this macro results, no? But why look at the object file and not `src/libcharon/sa/ikev2/tasks/ike_auth.c`?

**Comment by amodster** [comment]
Tobias
I did not find a call to METHOD() in ike_auth.c
Teledyne Confidential; Commercially Sensitive Business Data

**Reply by tobiasbrunner** [reply]
> I did not find a call to METHOD() in ike_auth.c

What do you mean?

**Comment by amodster** [comment]
ike_auth.c only had the definition of METHOD() function, I did not see a caller using METHOD()
I think it’s a macro of some kind
Teledyne Confidential; Commercially Sensitive Business Data

**Reply by tobiasbrunner** [reply]
It obviously is a macro, but that only defines the function. It has nothing to do with how it is called. The defined `build_i` function is assigned to the `task_t` interface and called when this particular task is executed by the IKEv2 task manager.
